### ISSUE TITLE
Specifying (or not) sync/async resolution behavior

### ISSUE THREAD

**Issue by briancavalier** [promises-aplus/promises-spec#4]
From #1

> Without getting into which is "better", I think the more important initial question is:
> 
> Does such a requirement actually belong in this proposal?
> 
> There are quality implementations of Promises/A in the wild that do each, and in my experience, they interoperate without problems. For example, I have mixed when.js, Dojo Deferred, jQuery Deferred, and Q in real, deployed applications over the past 2 years, and have never had a problem that I could attribute to one of them not doing async resolutions.
> 
> So, I think we need to answer the question of whether this belongs in the spec or not before we can discuss the obvious follow-on question :)

**Comment by maksimr** [comment]
Not clear to me the *pros* of always async promises.

Case when newbie may write something like:
```js
var value = "old";

promise.then(function () {
  value = "new";
});

console.log(value);
```
is really rare and easy detected case.

Problems with async behavior appear when you try writing sync tests:

```js
var value = "old";

// I want resolve promise immediately
promise.then(function () {
  value = "new";
});

assert.equal(value, 'new');
```

With current implementation it's very awkward. I should mock *setTimeout* function. Or write async tests with their disadvantages.

Thanks

**Comment by pkozlowski-opensource** [comment]
@maksimr what is needed in terms of testing is a way to simulate a "next tick" that a given promise implementation might use. For example $q service in AngularJS doesn't use setTimeout to trigger async resolution of promises.

Other than this I'm not sure what would be your proposal here - make all the promise implementations behave synchronously? 

Other though - you might want to look into https://github.com/domenic/chai-as-promised/ to see how a test-helper method might look like.

**Comment by maksimr** [comment]
@pkozlowski-opensource you are right. In angularjs we have mechanism like *$apply/$digest* which trigger *resolve/reject* of promises **synchronously**, but what if you don't use anuglarjs, how you should write tests?

If I correct understand source code of *chai-as-promised*, it is still **async tests**.
I want to do sync tests, because it is more robust, understandable(sync flow) and fast.

> Other than this I'm not sure what would be your proposal here - make all the promise implementations behave synchronously?

I want understand why promise A+ **always** should be asynchronous.

If the primary goal was `to eliminate a footgun` code like this:
```js
var value = 1;

promise.then(function () {
  value +=1;
});

console.log('expect', value, 'equal', 2);
```
Then it's a weak argument for me and this behavior only complicates testing now(because ` is needed in terms of testing is a way to simulate a "next tick" `).

**Comment by pkozlowski-opensource** [comment]
@maksimr what you are saying is not complete right - AnuglarJS still preserve asynchronous semantic (that is, execute in a different stack frame) while executing in the same JS turn. You seem to be confusing asynchronicity with a give JS context execution - those are not the same concepts.

Now - the most important part - what is your exact proposal? How would you make promises work asynchronously in the production code and synchronously in the test code? Does it even make sense?

**Comment by maksimr** [comment]
@pkozlowski-opensource 

>what is your exact proposal?

If I resolve promise immediately then it should call *then* handler immediately without `setTimeout`/`nextTick` in the same JS context.

**Comment by juandopazo** [comment]
Some reasons why: http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony

**Comment by petkaantonov** [comment]
FYI there is a concept of synchronous inspection implemented by promise libraries like bluebird and Q that is meant to solve this problem and more
